gps/deduce.go

```diff
@@ -571,6 +571,68 @@ func newDeductionCoordinator(superv *supervisor) *deductionCoordinator {
 	return dc
 }
 
+func newRegistryDeductionCoordinator(superv *supervisor, registry Registry) deducer {
+	dc := &registryDeductionCoordinator{
+		suprvsr:  superv,
+		rootxt:   radix.New(),
+		deducext: pathDeducerTrie(),
+		registry: registry,
+	}
+	return dc
+}
+
+type registryDeductionCoordinator struct {
+	suprvsr  *supervisor
+	mut      sync.RWMutex
+	rootxt   *radix.Tree
+	deducext *deducerTrie
+	registry Registry
+}
+
+func (dc *registryDeductionCoordinator) deduceRootPath(ctx context.Context, path string) (pathDeduction, error) {
+	if err := dc.suprvsr.ctx.Err(); err != nil {
+		return pathDeduction{}, err
+	}
+	// First, check the rootxt to see if there's a prefix match - if so, we
+	// can return that and move on.
+	dc.mut.RLock()
+	prefix, data, has := dc.rootxt.LongestPrefix(path)
+	dc.mut.RUnlock()
+	if has && isPathPrefixOrEqual(prefix, path) {
+		switch d := data.(type) {
+		case maybeSource:
+			return pathDeduction{root: prefix, mb: d}, nil
+		case *registryDeducer:
+			return d.deduce(ctx, path)
+		}
+		panic(fmt.Sprintf("unexpected %T in deductionCoordinator.rootxt: %v", data, data))
+	}
+
+	rd := &registryDeducer{
+		basePath: path,
+		registry: dc.registry,
+		suprvsr:  dc.suprvsr,
+		// The registry deducer will call this func with a completed
+		// pathDeduction if it succeeds in finding one. We process it
+		// back through the action channel to ensure serialized
+		// access to the rootxt map.
+		returnFunc: func(pd pathDeduction) {
+			dc.mut.Lock()
+			dc.rootxt.Insert(pd.root, pd.mb)
+			dc.mut.Unlock()
+		},
+	}
+
+	// Save the rd in the rootxt so that calls checking on similar
+	// paths made while the request is in flight can be folded together.
+	dc.mut.Lock()
+	dc.rootxt.Insert(path, rd)
+	dc.mut.Unlock()
+
+	// Trigger the HTTP-backed deduction process for this requestor.
+	return rd.deduce(ctx, path)
```

Comment:
This entails at least one HTTP request for each root path, including the known path case (which was removed above, compared with `deductionCoordinator.deduceRootPath()`). This will be a notable performance hit for the user vs. the current behavior.

More importantly, though, as long as we are guaranteeing that any proxy registry can access and passthrough-cache any upstream VCS source that a dep client would be able to access itself without an intermediate registry, then, then i see no compelling use case for having this extra request. In fact, all i see is a vector for confusion and errors, as we become reliant on registries to respect naming invariants that are presently universal, by virtue of being encoded in the tool. It seems like it should be possible - at least as a first pass approach - to just rely on the existing static deduction subsystem, then perform some structured transformations on the `pathDeduction`s that they return.

Is there a crucial use case that's serviced by making this request to the registry?

Reply:
Thanks. Yes, it would be preferable to deduct the root project path without HTTP request. I added the known paths deduction to address this issue.
 
However, for imports that are not using well known URLs, passthrough upstream VCS source is not always possible. For instance, in a closed network environment that work exclusively through a private registry deducing the project root by querying the VCS repository for the goget metadata cannot work (leaving aside the fact that most of the benefits of having a local private registry are lost by bypassing and accessing the sources directly).
Therefore, there is a valid use case IMO for having the registry know about the project roots it hosts. Once, we allowed static deduction to kick in first I don't think the performance impact (a single HEAD call) would be an issue.

Reply:
> Yes, it would be preferable to deduct the root project path without HTTP request.

great 😄

> However, for imports that are not using well known URLs,

for sure, i'm entirely in agreement that anything requiring normal `go-get` HTTP metadata deduction can be forwarded to the registry. 

> Once, we allowed static deduction to kick in first I don't think the performance impact (a single HEAD call) would be an issue.

agreed, that's no net increase in the number of HTTP requests, so no harm done.

Reply:
Great. Is it ready to be merged?